File: go/rerout/qr.go
// Package rerout — QR namespace.
//
// Reached via Client.QR(). Exposes a pure URL builder (URL) and an
// authenticated SVG fetcher (SVG).

package rerout

import (
	"context"
	"net/url"
	"sort"
	"strings"
)

// QR is the QR-helpers namespace. Construct via Client.QR().
type QR struct {
	client *Client
}

// URL builds the URL the Rerout API serves the QR SVG from. Pure — does not
// call the API. Authentication is the caller's responsibility, since plain
// <img src=…> tags can't add a bearer token; typically you pass the URL
// through a server-side proxy.
//
//	u := c.QR().URL("q4", nil)
//	// → https://api.rerout.co/v1/links/q4/qr
//
//	branded := c.QR().URL("q4", &rerout.QROptions{
//	    Size:   rerout.Int(12),
//	    ECC:    "H",
//	    Domain: "go.brand.com",
//	})
func (q *QR) URL(code string, options *QROptions) string {
	return BuildQRURL(q.client.BaseURL(), code, options)
}

// SVG fetches the rendered QR as an SVG string. Hits the same endpoint as
// URL but attaches the bearer token and returns the body.
func (q *QR) SVG(ctx context.Context, code string, options *QROptions) (string, error) {
	q2 := qrQueryValues(options)
	var out string
	err := q.client.do(ctx, requestOptions{
		method:      "GET",
		path:        "/v1/links/" + url.PathEscape(code) + "/qr",
		query:       q2,
		rawResponse: true,
	}, &out)
	if err != nil {
		return "", err
	}
	return out, nil
}

// BuildQRURL is the standalone version of QR.URL. Useful when you have a
// base URL handy but not a full Client (e.g. server-side template helpers).
//
// Trailing slashes on baseURL are stripped; the code is path-escaped.
func BuildQRURL(baseURL, code string, options *QROptions) string {
	base := trimTrailingSlashes(baseURL)
	path := base + "/v1/links/" + url.PathEscape(code) + "/qr"
	values := qrQueryValues(options)
	if len(values) == 0 {
		return path
	}
	// Use a deterministic ordering so the function is pure & testable.
	keys := make([]string, 0, len(values))
	for k := range values {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	var sb strings.Builder
	sb.WriteString(path)
	sb.WriteByte('?')
	for i, k := range keys {
		if i > 0 {
			sb.WriteByte('&')
		}
		sb.WriteString(url.QueryEscape(k))
		sb.WriteByte('=')
		sb.WriteString(url.QueryEscape(values.Get(k)))
	}
	return sb.String()
}

func qrQueryValues(options *QROptions) url.Values {
	q := options.toQueryParameters()
	if len(q) == 0 {
		return nil
	}
	out := make(url.Values, len(q))
	for k, v := range q {
		out.Set(k, v)
	}
	return out
}

File: ruby/lib/rerout/qr.rb
# frozen_string_literal: true

require 'erb'
require 'uri'

module Rerout
  module Resources
    # QR helpers.
    class Qr
      # @param client [Rerout::Client]
      def initialize(client)
        @client = client
      end

      # Build the URL the API serves the QR SVG from. Pure — does not hit
      # the network.
      #
      # @param code [String]
      # @param options [Rerout::QrOptions, Hash, nil]
      # @return [String]
      def url(code, options = nil)
        Qr.build_url(base_url: @client.base_url, code: code, options: options)
      end

      # Fetch the QR as an SVG string. Attaches the bearer token.
      #
      # @param code [String]
      # @param options [Rerout::QrOptions, Hash, nil]
      # @return [String] the rendered SVG markup.
      def svg(code, options = nil)
        raise ArgumentError, 'code is required' if code.nil? || code.to_s.empty?

        qopts = Qr.coerce_options(options)
        path = "/v1/links/#{ERB::Util.url_encode(code.to_s)}/qr"
        @client.request(
          method: :get,
          path: path,
          query: qopts.to_query_hash,
          raw: true
        )
      end

      # @api private — also used by callers that just want a URL without a client.
      def self.build_url(base_url:, code:, options: nil)
        raise ArgumentError, 'code is required' if code.nil? || code.to_s.empty?

        qopts = coerce_options(options)
        base = base_url.to_s.sub(%r{/+\z}, '')
        path = "/v1/links/#{ERB::Util.url_encode(code.to_s)}/qr"
        url = "#{base}#{path}"
        query = qopts.to_query_hash
        return url if query.empty?

        "#{url}?#{URI.encode_www_form(query)}"
      end

      # @api private
      def self.coerce_options(options)
        case options
        when nil then QrOptions.new
        when QrOptions then options
        when Hash then QrOptions.new(**options.transform_keys(&:to_sym))
        else
          raise ArgumentError, 'options must be a Rerout::QrOptions, Hash, or nil'
        end
      end
    end
  end
end
